covid_graphs/covid_graphs/country_graph.py

```diff
@@ -44,59 +64,75 @@ def __init__(
         self.short_name = report.short_name
         self.long_name = report.long_name
 
-        # TODO(lukas): Figure out better strategy for more predictions
         if len(country_predictions) >= 1:
-            self.prediction_date = country_predictions[0].prediction_event.date
+            self.prediction_date = max(
+                country_prediction.prediction_event.date
+                for country_prediction in country_predictions
+            )
 
-        self.curves = [
-            prediction.formula.get_curve(country_report=report)
+        # We create traces in three steps:
+        # 1. A formula is first shifted to the appropriate date on the graph, creating a
+        #    trace generator.
+        # 2. The trace generators are used to calculate the last date of the graph, since each trace
+        #    generator stores the minimal display length of its trace.
+        # 3. Once we know the date range of the graph, we can plot the formulas, creating traces.
+        trace_generators = [
+            prediction.formula.get_trace_generator(country_report=report)
             for prediction in country_predictions
         ]
+        start_date, display_until = _get_display_range(report, trace_generators)
+
+        self.traces = [
+            _create_trace(trace_generator, display_until) for trace_generator in trace_generators
+        ]
 
-        min_start_date, self.display_until = _get_display_range(report, self.curves)
-        min_start_date_idx = report.dates.index(min_start_date)
+        start_date_idx = report.dates.index(start_date)
         # Crop country data to display.
-        self.cropped_dates = report.dates[min_start_date_idx:]
-        self.cropped_cumulative_active = report.cumulative_active[min_start_date_idx:]
+        self.cropped_dates = report.dates[start_date_idx:]
+        self.cropped_cumulative_active = report.cumulative_active[start_date_idx:]
+        self.log_xaxis_date_since = self.cropped_dates[0] - datetime.timedelta(days=1)
+        self.log_title = f"Days [since {self.log_xaxis_date_since.strftime('%b %d, %Y')}]"
+        self.date_title = f"Date [starting from {self.cropped_dates[0].strftime('%b %d, %Y')}]"
```

Comment:
I'd suggest `self.date_title = "Date"`, "starting from" does not make really sense.

Reply:
Done (for context, we used to have number of cases in the sentence)